src/shared/dissect-image.c

```diff
@@ -1289,8 +1295,37 @@ static int verity_partition(
 #endif
         } else
                 r = crypt_activate_by_volume_key(cd, name, root_hash, root_hash_size, CRYPT_ACTIVATE_READONLY);
-        if (r < 0)
+        if (r < 0 && r != -EEXIST)
                 return r;
+        if (r == -EEXIST) {
+                /* If the same volume was already open, check that the root hashes match, and reuse it if they do */
+                _cleanup_free_ char *root_hash_existing = NULL;
+                struct crypt_params_verity crypt_params = {};
+                size_t root_hash_existing_size = root_hash_size;
+
+                crypt_free(cd);
```

Comment:
you need to set `cd` to NULL here. Otherwsie, if `crypt_init_by_name()` in the next line fails, `cd` will point to freed memory and we'd destroy freed memory again...

Reply:
fixed - now this logic live in its own static helper function, with its own crypt_device